### PR TITLE
Add EdgeCraftRag as a GenAIExample

### PR DESCRIPTION
**Pull request by Yongbozzz**
## Description

Edge Craft RAG (EC-RAG) is a customizable, tunable and production-ready Retrieval-Augmented Generation system for edge solutions. It is designed to curate the RAG pipeline to meet hardware requirements at edge with guaranteed quality and performance.

## Issues

N/A

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds new functionality)

## Dependencies

langchain-core, llama-index

## Tests

image

## Related RFC

[https://opea-project.github.io/latest/community/rfcs/24-08-21-GenAIExample-002-Edge_Craft_RAG.html](https://opea-project.github.io/latest/community/rfcs/24-08-21-GenAIExample-002-Edge_Craft_RAG.html)

### CODE REVIEW

**Review by ftian1 — COMMENTED** [review]
looks like the edgecraftrag folder contains an entire solution of rag server, from the component to the pipeline construction, which is overlapped with OPEA design. OPEA design philosophy is trying to abstract those common component for enterprise use and the pipeline construction is done in OPEA mega service. I am not sure if there are values to introduce another pipeline implementation.  and you also invokes GenAIComps.core.register_microservice() API in your implementation, which shouldn't happen. we only register OPEA microservice for those common components and should put under GenAIComps repo but not here.

**Comment by myqi** [comment]
> looks like the edgecraftrag folder contains an entire solution of rag server, from the component to the pipeline construction, which is overlapped with OPEA design. OPEA design philosophy is trying to abstract those common component for enterprise use and the pipeline construction is done in OPEA mega service. I am not sure if there are values to introduce another pipeline implementation. and you also invokes GenAIComps.core.register_microservice() API in your implementation, which shouldn't happen. we only register OPEA microservice for those common components and should put under GenAIComps repo but not here.

@ftian1 Thanks for your comment, as mentioned in the [RFC](https://github.com/myqi/opea-docs/blob/ec-rag/community/rfcs/24-08-21-GenAIExample-002-Edge_Craft_RAG.md), the goal of this example is to make a RAG solution specifically for edge use cases, which includes these characteristics:
- Less microservices are needed in a single node on-prem system.
- Intel Core CPU, NPU, iGPU, dGPU will be selectively used in a single pipeline.

Thus, we are building the pipeline as a solution for now to 1) eliminate some of the microservices setup 2) implement each of the stages in the pipeline to meet edge specific needs. Moving on, we are happy to contribute these implementations to GenAIComps. 

We are showcasing that OPEA can integrate with existing solutions by just replacing some of the components with OPEA microservices (Like LLM in this case). With that, I don't think it's an overlap for the implementation. We are NOT trying to change the way of building the pipeline in OPEA.

As to the register_microservice() API, we're okay to use another API serving framework and will update soon.

**Comment by lvliang-intel** [comment]
@Yongbozzz,
Please fix the code spell issue.

**Comment by myqi** [comment]
> @Yongbozzz,
> Please fix the code spell issue.
> Uploading image.png…

EdgeCraftRAG/edgecraftrag/api_schema.py:9: ModelIn ==> modeling, model in

ModelIn is a type, can this be ignored?

**Review by ZePan110 — APPROVED** [review]

File: EdgeCraftRAG/tests/test_pipeline_local_llm.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
+{
```
Comment (chensuyue) [inline EdgeCraftRAG/tests/test_pipeline_local_llm.json:1]: We need e2e test pass before merge.

File: EdgeCraftRAG/Dockerfile.ui [hunk 2]
```diff
@@ -0,0 +1,23 @@
+FROM python:3.11-slim
```
Comment (chensuyue) [inline EdgeCraftRAG/Dockerfile.ui:1]: Align with other code structure, the ui dockerfile should be under, `ui/docker/xxx`

File: EdgeCraftRAG/docker_compose/Intel/Arc/compose.yaml [hunk 3]
```diff
@@ -0,0 +1,78 @@
+# Copyright (C) 2024 Intel Corporation
```
Comment (chensuyue) [inline EdgeCraftRAG/docker_compose/Intel/Arc/compose.yaml:1]: `Intel/Arc/compose.yaml` -> `intel/gpu/arc/compose.yaml`

**Comment by ftian1** [comment]
I am ok with the overall design. pls fix the comments provided by suyue

File: EdgeCraftRAG/docker_compose/Intel/Arc/compose.yaml (hunk 3 above)
Comment (Yongbozzz) [reply EdgeCraftRAG/docker_compose/Intel/Arc/compose.yaml]: resolved

File: EdgeCraftRAG/Dockerfile.ui (hunk 2 above)
Comment (Yongbozzz) [reply EdgeCraftRAG/Dockerfile.ui]: resolved

File: EdgeCraftRAG/tests/test_pipeline_local_llm.json (hunk 1 above)
Comment (Yongbozzz) [reply EdgeCraftRAG/tests/test_pipeline_local_llm.json:1]: Hi suyue ,as we aligned , I will add e2e test later after merge,,I will do it ASAP. thanks

**Comment by Yongbozzz** [comment]
> I am ok with the overall design. pls fix the comments provided by suyue

Hi Tian Feng, I have resolve the comments provided by suyue and will add e2e test after merge, thanks

**Review by lvliang-intel — APPROVED** [review]

**Pull request merged**